### PR TITLE
New job chat link preview

### PR DESCRIPTION
**Pull request by samschlicht**
I re-submitted this because the AWS code guru was referring to some code that I have since changed. It may make similar comments again, but leaving the old ones in was prone to cause confusion.

Reminder that the vid I made is in the ZH issue (it's too big to upload here).

### CODE REVIEW

File: server/src/main/java/org/tctalent/server/service/db/impl/LinkPreviewServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+
+    } catch (HttpStatusException e) {
+      // If doc can't be retrieved, log the error and return null.
+
+      LogBuilder.builder(log)
+          .action("BuildLinkPreview: Jsoup.connect(url).get()")
+          .message(e.getMessage())
+          .logInfo();
+
+      return null;
```
Comment (sadatmalik) [inline server/src/main/java/org/tctalent/server/service/db/impl/LinkPreviewServiceImpl.java:98]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception has been ignored. The catch-block hides information about the stack trace. Not logging or forwarding the stack trace of an exception can mask unexpected errors and complicated debugging.

**Fix**: Use a catch block to handle the exception and write a log message declaring that the exception happened. 
[Learn more.](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)

Comment (samschlicht) [reply server/src/main/java/org/tctalent/server/service/db/impl/LinkPreviewServiceImpl.java:98]: Taken care of this one.

File: server/src/main/java/org/tctalent/server/model/db/LinkPreview.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import lombok.Getter;
+import lombok.Setter;
+import org.springframework.lang.Nullable;
+
+@Getter
+@Setter
+@Entity
+@Table(name = "link_preview")
+@SequenceGenerator(name = "seq_gen", sequenceName = "link_preview_id_seq", allocationSize = 1)
+public class LinkPreview extends AbstractDomainObject<Long> {
```
Comment (camerojo) [inline server/src/main/java/org/tctalent/server/model/db/LinkPreview.java:34]: Some JavaDoc would be good - at a class level explaining what a LinkPreview actually is, and also JavaDoc for each field.

Comment (camerojo) [reply server/src/main/java/org/tctalent/server/model/db/LinkPreview.java:34]: Similarly, unrelated to link previews, I note that there is no JavaDoc for Reaction - could you add some doc to the Reaction class describing briefly what a Reaction is. Reaction is a pretty generic term, and someone coming cold to the code could find it useful to have some doc noting that in this context a Reaction is just an icon - like a thumbs up etc.

File: server/src/main/java/org/tctalent/server/service/db/LinkPreviewService.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+/**
+ * Implements Jsoup web-scraping and HTML parsing to return link previews.
+ * Uses <a href="https://andrejgajdos.com/how-to-create-a-link-preview/">...</a> as a starting point
+ * for selecting content from the many divergent options available across different websites.
+ *
+ * @author samschlicht
+ */
+
+public interface LinkPreviewService {
+
```
Comment (camerojo) [inline server/src/main/java/org/tctalent/server/service/db/LinkPreviewService.java:30]: Nice doc on this service - so we are just missing a bit of doc on the entity

File: server/src/main/java/org/tctalent/server/service/db/impl/LinkPreviewServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+import org.tctalent.server.exception.InvalidRequestException;
+import org.tctalent.server.logging.LogBuilder;
+import org.tctalent.server.model.db.LinkPreview;
+import org.tctalent.server.repository.db.LinkPreviewRepository;
+import org.tctalent.server.service.db.LinkPreviewService;
+
+@Slf4j
+@RequiredArgsConstructor
+@Service
+public class LinkPreviewServiceImpl implements LinkPreviewService {
```
Comment (camerojo) [inline server/src/main/java/org/tctalent/server/service/db/impl/LinkPreviewServiceImpl.java:43]: A lot of the logic in here is hard to follow without a broad understanding how previews are generated - a link to a doc like https://andrejgajdos.com/how-to-create-a-link-preview/ would be helpful to any developer coming into this code cold and trying to add or fix something.

**Review by camerojo — APPROVED** [review]
Really great work!! I have just asked for a little more internal doc - see comments.

But I have approved nevertheless as I know that you are keen to see this working in staging.

File: server/src/main/java/org/tctalent/server/service/db/impl/LinkPreviewServiceImpl.java (hunk 4 above)
Comment (samschlicht) [reply server/src/main/java/org/tctalent/server/service/db/impl/LinkPreviewServiceImpl.java:43]: @camerojo I actually already linked to that article and gave a little doc overview in the service interface: should I add some to the implementation, too?

Absolutely agree on the entities — I'll merge this and make a separate issue to add some doc.
